### ISSUE TITLE
Weird TypeAnalysis circular dependency?

### ISSUE THREAD

**Issue by TomasPuverle** [souffle-lang/souffle#1826]
In `ast::TranslationUnit::getAnalysis<>`, there is a weird dependency when using `analysis::TypeAnalysis`.  For example, if the following code

```
 analyses[name] = mk<Analysis>();
 analyses[name]->run(*this);
```

is changed to

```
auto ana = mk<Analysis>();
ana->run(*this);
analysis[name] = std::move(ana);
```

this will result in a SEGV due to stack overflow.  Here's the relevant part of the stack:

```
#58 0x000000000053acb7 in souffle::ast::analysis::TypeAnalysis::analyseTypes (
    tu=..., clause=..., logs=0x0) at ast/analysis/Type.cpp:127
#59 0x000000000053cc37 in souffle::ast::analysis::TypeAnalysis::run (
    this=0xa42c10, translationUnit=...) at ast/analysis/Type.cpp:461
#60 0x000000000052941c in souffle::ast::TranslationUnit::addAnalysis (
    this=0x69c160, name=0x5ae6c8 "type-analysis", 
    analysis=std::unique_ptr<souffle::ast::analysis::Analysis> = {...})
    at ast/TranslationUnit.cpp:31
#61 0x000000000052af84 in souffle::ast::TranslationUnit::getAnalysis<souffle::ast::analysis::TypeAnalysis> (this=0x69c160) at ./ast/TranslationUnit.h:57
#62 0x0000000000541339 in souffle::ast::analysis::TypeConstraintsAnalysis::TypeConstraintsAnalysis (this=0x7fffff804790, tu=...)
    at ./ast/analysis/TypeConstraints.h:85
#63 0x000000000053acb7 in souffle::ast::analysis::TypeAnalysis::analyseTypes (
    tu=..., clause=..., logs=0x0) at ast/analysis/Type.cpp:127
```

(Note that the `addAnalysis` is not in master, and it just a small helper I added to print the debugging into, which does not affect the code flow).  

Basically, I was trying to avoid adding the analysis to the map until it has successfully completed (in case there was an error/exception etc).

**Comment by azreika** [comment]
Could be related to issue #1739.

**Comment by TomasPuverle** [comment]
I don't think they are related - #1739 leads to an infinite non-converging type analysis loop.  This problem leads to a crash and stack overflow.
AFAICS, the problem is that TypeAnalysis depends on TypeAnalysis.  The "loop breaking" behaviour is the presence of the type analysis object in the analysis map in `ast::TranslationUnit`, even though it has not been run yet.  
I mostly logged this because this could indicate an circular issue in the design.
If this is intentional, I'm happy to close.

**Comment by azreika** [comment]
There should be no cyclic loops in any analyses, so if there's a circular design issue it is not intentional.

**Comment by TomasPuverle** [comment]
Ok, in which case: 

TypeAnalysis -> 
depends on TypeConstraintAnalysis (Type.cpp:127 calls `    return TypeConstraintsAnalysis(tu).analyse(clause, logs);`) ->
depends on TypeAnalysis (by using it as a member, at TypeConstraint.h:91, `    const TypeAnalysis& typeAnalysis = *tu.getAnalysis<TypeAnalysis>();`)
